src/org/python/core/ThreadState.java

```diff
@@ -43,14 +43,17 @@ public PySystemState getSystemState() {
         PySystemState systemState = systemStateRef == null ? null : systemStateRef.get();
         return systemState == null ? Py.defaultSystemState : systemState; 
     }
-    
+
     public boolean enterRepr(PyObject obj) {
         if (reprStack == null) {
             reprStack = new PyList(new PyObject[] {obj});
             return true;
         }
+        Boolean objHasTp = (obj instanceof Traverseproc);
         for (int i = reprStack.size() - 1; i >= 0; i--) {
-            if (obj == reprStack.pyget(i)) {
+            if (objHasTp && obj == reprStack.pyget(i)){
```

Comment:
I don't follow this. Looks plausible but a comment from @Stewori would reassure me. Again, is this maybe a thing that was lacking all along, not a consequence of this change?

Reply:
Sorry I don't really see the relation between ``Traverseproc`` and ``JavaProxy``.
@devendor could you please elaborate what you intend to do here?
Why is checking the JavaProxy relevant if no Traverseproc is implemented?